e2e/engine/cmd.go

```diff
@@ -285,8 +253,62 @@ func (e *engineCmd) Run(ctx context.Context, eventC chan *node.Event) error {
 				}
 			}
 		}
+		if e.conf.CheckAppHashAfterCmd {
+			if err := checkapphash(e.conf.Nodes); err != nil {
+				return errors.Wrapf(err, "check apphash failed after test command, %s", n.RunCmd)
+			}
+		}
+	}
+
+	return nil
+}
+
+func checkapphash(nodes map[string]*node.Node) error {
+	time.Sleep(time.Second * 1)
+	fmt.Printf("--> run all: %v \n", "checkapphash")
+	var apphash = make(map[string]struct{})
+	var lastBlockHeight int64
+	for _, v := range nodes {
+		u := fmt.Sprintf("%s/abci_info", v.RPCAddress)
+		resp, err := http.Get(u)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 {
+			respBytes, _ := ioutil.ReadAll(resp.Body)
+			return fmt.Errorf("post status not OK: %s, response body: %s", resp.Status, string(respBytes))
+		}
+		var info = struct {
+			JSONRPC string `json:"jsonrpc"`
+			ID      string `json:"id"`
+			Result  struct {
+				Response abciResponseInfo2 `json:"response"`
+			} `json:"result"`
+		}{}
+
+		err = json.NewDecoder(resp.Body).Decode(&info)
+		if err != nil {
+			return err
+		}
+		newLastBlockHeight, err := strconv.ParseInt(info.Result.Response.LastBlockHeight, 10, 64)
+		if err != nil {
+			return err
+		}
+		if lastBlockHeight == 0 {
+			lastBlockHeight = newLastBlockHeight
+		}
+		if lastBlockHeight == newLastBlockHeight {
+			apphash[string(info.Result.Response.LastBlockAppHash)] = struct{}{}
+			fmt.Printf("--> GET: %s, AppHash: %0xX\n", u, info.Result.Response.LastBlockAppHash)
+		}
 	}
 
+	// apphash should have only 1 entry
+	// this might not be true if network latency is hight
+	if len(apphash) != 1 {
+		return fmt.Errorf("Wrong Block.Header.AppHash")
+	}
 	return nil
```

Comment:
Should this function check that all app hashes are the same before returning nil?

Reply:
apphash is a map, so `len(apphash) == 1` confirms there is at least one apphash and that they are all the same.